ember-core/src/test/scala/org/http4s/ember/core/StreamingParserSuite.scala

```diff
@@ -0,0 +1,198 @@
+import org.http4s.Http4sSuite
+import org.scalacheck.Gen
+import org.scalacheck.Prop._
+import org.scalacheck.effect.PropF
+
+class StreamingParserSuite extends Http4sSuite {
+
+  object Helpers {
+    def taking[F[_]: Concurrent, A](segments: List[List[A]]): F[F[Option[Chunk[A]]]] =
+      for {
+        q <- Queue.unbounded[F, Option[Chunk[A]]]
+        _ <- segments.traverse(bytes => q.enqueue1(Some(Chunk.seq(bytes))))
+        _ <- q.enqueue1(None)
+      } yield q.dequeue1
+
+    def subdivided[A](as: List[A], count: Int): Gen[List[List[A]]] = {
+      def go(out: List[List[A]], remaining: Int): Gen[List[List[A]]] =
+        if (remaining <= 0) Gen.const(out)
+        else {
+          Gen.chooseNum(0, out.length - 1).flatMap { idx =>
+            val prefix = out.take(idx)
+            val curr = out(idx)
+            val suffix = out.drop(idx + 1)
+
+            Gen.chooseNum(0, curr.length - 1).flatMap { splitIdx =>
+              val (l, r) = curr.splitAt(splitIdx)
+              val split = List(l, r).filterNot(_.isEmpty)
+              val next = List(prefix, split, suffix).filterNot(_.isEmpty).flatten
+              go(next, remaining - 1)
+            }
+          }
+        }
+
+      if (as.isEmpty) Gen.const(Nil) else Gen.delay(go(List(as), count))
+    }
+
+    def debug(segments: List[List[Byte]]): Unit = {
+      // handleErrorWith(e => {e.printStackTrace();IO(false)}).assert
+
+      println("----")
+      segments.foreach { s =>
+        println(s)
+      }
+      // println("------")
+      // segments.foreach { s =>
+      //   s.foreach { byte =>
+      //     if (byte == 13) print("CR")
+      //     else if (byte == 10) print("NL")
+      //     else print(new String(Array[Byte](byte)))
+      //   }
+      //   println()
+      // }
+    }
+  }
+
+  object Fixtures {
+    val RequestFixedBody = toBytes(
+      List(
+        "POST /foo HTTP/1.1\r\n",
+        "Content-Type: text/plain\r\n",
+        "Content-Length: 5\r\n\r\n",
+        "hello"
+      ))
+
+    val RequestChunkedBody = toBytes(
+      List(
+        "POST /foo HTTP/1.1\r\n",
+        "Content-Type: text/plain\r\n",
+        "Transfer-Encoding: chunked\r\n\r\n",
+        "7\r\n",
+        "Mozilla\r\n",
+        "9\r\n",
+        "Developer\r\n",
+        "7\r\n",
+        "Network\r\n",
+        "0\r\n",
+        "\r\n"
+      ))
+
+    val ResponseFixedBody = toBytes(
+      List(
+        "HTTP/1.1 200 OK\r\n",
+        "Content-Type: text/plain\r\n",
+        "Content-Length: 5\r\n\r\n",
+        "hello"
+      ))
+
+    val ResponseVariableBody = toBytes(
+      List(
+        "HTTP/1.1 200 OK\r\n",
+        "Content-Type: text/plain\r\n",
+        "Transfer-Encoding: chunked\r\n\r\n",
+        "7\r\n",
+        "Mozilla\r\n",
+        "9\r\n",
+        "Developer\r\n",
+        "7\r\n",
+        "Network\r\n",
+        "0\r\n",
+        "\r\n"
+      ))
+
+    // TODO: Express in terms of Gen[Request[IO]] and Gen[Response[IO]]
```

Comment:
This is a nice-to-have

Reply:
You mean to generate from arbitrary requests and responses to `List[List[Byte]]`?

Reply:
Yeah exactly. Though per our Gitter conversation earlier, I'm interested in generating raw HTTP messages and then chunking that